Common/include/basic_types/ad_structure.hpp

```diff
@@ -522,6 +539,39 @@ namespace AD{
 
   FORCEINLINE void EndPassive(bool wasActive) { if(wasActive) StartRecording(); }
 
+  FORCEINLINE bool PausePreaccumulation() {
+    const auto current = PreaccEnabled;
+    if (!current) return false;
+    SU2_OMP_BARRIER
+    SU2_OMP_MASTER
+    PreaccEnabled = false;
+    END_SU2_OMP_MASTER
+    SU2_OMP_BARRIER
+    return true;
+  }
+
+  FORCEINLINE void ResumePreaccumulation(bool wasActive) {
+    if (!wasActive) return;
+    SU2_OMP_BARRIER
+    SU2_OMP_MASTER
+    PreaccEnabled = true;
+    END_SU2_OMP_MASTER
+    SU2_OMP_BARRIER
+  }
+
+  FORCEINLINE void StartNoSharedReading() {
+#ifdef HAVE_OPDI
+    opdi::logic->setAdjointAccessMode(opdi::LogicInterface::AdjointAccessMode::Classical);
+    opdi::logic->addReverseBarrier();
```

Comment:
Thanks for this, I think when I tried it I forgot the reverse barrier.
If performance testing goes well this mode can also be enabled in the `::Source_Residual` routines of the solvers.

Reply:
Not for all parts though, can it? `CEulerSolver::SourceResidual`, for example, has a computational path that leads to `computeGradientsGreenGauss` where preacc had to be disabled due to shared reading.

Reply:
Only for the iPoint loops involving numerics where we "Set" data based on iPoint alone.

Reply:
I agree. I added this for the `Source_Residual` routines that use OpenMP in 2830dea1c0da1cc6d6a53bf0a3403723de5d2ba7.